src/nnet3/nnet-common.cc

```diff
@@ -151,7 +152,175 @@ void ReadIndexVector(std::istream &is, bool binary,
   }
 }
 
+static void WriteCindexVectorElementBinary(
+    std::ostream &os,
+    const std::vector<Cindex> &vec,
+    int32 i) {
+  bool binary = true;
+  int32 node_index = vec[i].first;
+  Index index = vec[i].second;
+  if (i == 0 || node_index != vec[i-1].first) {
+    // '|' into ranges that each have all the same node name, like:
+    // [node_1: index_1 index_2] [node_2: index_3 index_4]
+    os.put('|');
+    WriteBasicType(os, binary, node_index);
+  }  
+  if (i == 0) {
+    if (index.n == 0 && index.x == 0 &&
+        std::abs(index.t) < 125) {
+      // handle this common case in one character.
+      os.put(static_cast<signed char>(index.t));
+    } else {  // handle the general case less efficiently.
+      os.put(127);
+      WriteBasicType(os, binary, index.n);
+      WriteBasicType(os, binary, index.t);
+      WriteBasicType(os, binary, index.x);
+    }
+  } else {
+    Index last_index = vec[i-1].second;
+    if (index.n == last_index.n && index.x == last_index.x &&
+        std::abs(index.t - last_index.t) < 125) {
+      signed char c = index.t - last_index.t;
+      os.put(c);
+    } else {  // handle the general case less efficiently.
+      os.put(127);
+      WriteBasicType(os, binary, index.n);
+      WriteBasicType(os, binary, index.t);
+      WriteBasicType(os, binary, index.x);
+    }
+  }
+  if (i == vec.size())
+    os << ']';
+  if (!os.good())
+    KALDI_ERR << "Output stream error detected";
+}
+
+static void ReadCindexVectorElementBinary(
+    std::istream &is,
+    int32 i,
+    std::vector<Cindex> *vec) {
+  bool binary = true;
+  Index &index = (*vec)[i].second;
+  if (!is.good())
+    KALDI_ERR << "End of file while reading vector of Cindex.";
+  if (is.peek() == static_cast<int>('|')) {
+    is.get();
+    ReadBasicType(is, binary, &((*vec)[i].first));
+  } else {
+    (*vec)[i].first = (*vec)[i-1].first;
+  }
+  signed char c = is.get();
+  if (i == 0) {
+    if (std::abs(int(c)) < 125) {
+      index.n = 0;
+      index.t = c;
+      index.x = 0;
+    } else {
+      if (c != 127)
+        KALDI_ERR << "Unexpected character " << c
+                  << " encountered while reading Cindex vector.";
+      ReadBasicType(is, binary, &(index.n));
+      ReadBasicType(is, binary, &(index.t));
+      ReadBasicType(is, binary, &(index.x));
+    }
+  } else {
+    Index &last_index = (*vec)[i-1].second;
+    if (std::abs(int(c)) < 125) {
```

Comment:
The way you are writing these things is very inefficient in the normal case- it gives almost no compression.
Remember in the normal computation, indexes are first sorted on t and then n.  So indexes differing only in n is actually the normal case (and they'll normally differ only by 1).  However, it's possible that some components may sort their input differently and sort instead on n first, so indexes differing only in n should also be supported to be written efficiently.  I suggest to allocate two different ranges of the character for each case (differing only in n vs. differing only in t).